hardware/kubernetes_client.go

```diff
@@ -0,0 +1,255 @@
+// See k8s.io/client-go/tools/clientcmd for constructing *rest.Config objects.
+func NewKubernetesClient(config KubernetesClientConfig) (*KubernetesClient, error) {
+	opts := tink.GetServerOptions()
+	opts.Namespace = config.Namespace
+
+	// Use a manager from the tink project so we can take advantage of the indexes and caching it configures.
+	// Once started, we don't really need any of the manager capabilities hence we don't store it in the
+	// KubernetesClient
+	manager, err := tink.NewManager(config.Config, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	managerCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		if err := manager.Start(managerCtx); err != nil {
+			panic(err)
+		}
+	}()
+
+	client := NewKubernetesClientWithClient(manager.GetClient())
+	client.close = cancel
+	client.waitForCacheSync = manager.GetCache().WaitForCacheSync
+
+	return client, nil
+}
+
+// ListerClient lists Kubernetes resources using a sigs.k8s.io/controller-runtime client.
+type ListerClient interface {
+	List(ctx context.Context, list crclient.ObjectList, opts ...crclient.ListOption) error
+}
+
+// NewKubernetesClientWithClient creates a new KubernetesClient instance that uses client to find resources. The
+// Close() and WaitForCacheSync() methods of the returned client are noops.
+func NewKubernetesClientWithClient(client ListerClient) *KubernetesClient {
+	return &KubernetesClient{
+		client:           client,
+		close:            func() {},
+		closeM:           &sync.RWMutex{},
+		waitForCacheSync: func(context.Context) bool { return true },
+	}
+}
+
+// Close stops synchronization with the cluster. Subsequent calls to all of k's methods, excluding IsHealthy(), have
+// undefined behavior.
+func (k *KubernetesClient) Close() {
+	k.closeM.Lock()
+	defer k.closeM.Unlock()
+
+	// nil indicates Close() has been called for other methods.
+	k.close()
+	k.close = nil
+}
+
+// WaitForCacheSync waits for the internal client cache to synchronize.
+func (k *KubernetesClient) WaitForCacheSync(ctx context.Context) bool {
+	return k.waitForCacheSync(ctx)
+}
+
+// IsHealthy returns true until k.Close() is called. It is thread safe.
+func (k *KubernetesClient) IsHealthy(context.Context) bool {
+	k.closeM.RLock()
+	defer k.closeM.RUnlock()
+	return k.close != nil
+}
+
+// ByIP retrieves a hardware resource associated with ip.
+func (k *KubernetesClient) ByIP(ctx context.Context, ip string) (Hardware, error) {
+	var hw tinkv1alpha1.HardwareList
+	err := k.client.List(ctx, &hw, crclient.MatchingFields{
+		tink.HardwareIPAddrIndex: ip,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hw.Items) == 0 {
+		return nil, fmt.Errorf("no hardware with ip '%v'", ip)
+	}
+
+	if len(hw.Items) > 1 {
+		names := make([]string, len(hw.Items))
+		for i, item := range hw.Items {
+			names[i] = item.Name
+		}
+		return nil, fmt.Errorf("multiple hardware with ip '%v': [%v]", ip, strings.Join(names, ", "))
```

Comment:
This seems like an information leak, maybe don't return the enumerated list? Log it for an admin that cares, but the client doesn't need to know

Reply:
This is from the other PR, will update there.